### PR TITLE
Place Roslyn under a separate directory in MSBuild.

### LINKED ISSUE #14660: Roslyn should move into a subdirectory of MSBuild in install
Our current location, next to the MSBuild binaries, is causing problems due to loading installed Roslyn binaries when we mean to load binaries from NuGet or the bootstrap build.

We should instead move the Roslyn compilers to a separate directory so we will not interfere with other Roslyn loads.

### PR DESCRIPTION
**Pull request by OmarTawfik**
@dotnet/roslyn-compiler @dotnet/roslyn-infrastructure @rainersigwald 
Fixes #14660

### CODE REVIEW

File: src/Compilers/Core/MSBuildTask/Utilities.cs [hunk 1]
```diff
@@ -149,5 +151,23 @@ internal static string GetLocation(Assembly assembly)
 
             return (string)method.Invoke(assembly, parameters: null);
         }
+
+        internal static string GenerateFullPathToMSBuildRoslynTool(string toolName)
+        {
+            // Roslyn always deploys to the 32Bit folder of MSBuid, so request this path on all architectures.
```
Comment (rainersigwald) [inline src/Compilers/Core/MSBuildTask/Utilities.cs]: Typo: `MSBuid` -> `MSBuild`

File: src/Setup/DevDivVsix/CompilersPackage/Microsoft.CodeAnalysis.Compilers.swr [hunk 2]
```diff
@@ -6,61 +6,7 @@ package name=Microsoft.CodeAnalysis.Compilers
...
-    file source=$(OutputPath)\Vsix\CompilerExtension\System.Security.Cryptography.X509Certificates.dll vs.file.ngenArchitecture=all
-    file source=$(OutputPath)\Vsix\CompilerExtension\System.Security.Principal.Windows.dll vs.file.ngenArchitecture=all
-    file source=$(OutputPath)\Vsix\CompilerExtension\System.Text.Encoding.CodePages.dll vs.file.ngenArchitecture=all
-    file source=$(OutputPath)\Vsix\CompilerExtension\System.Threading.Thread.dll vs.file.ngenArchitecture=all
-    file source=$(OutputPath)\Vsix\CompilerExtension\System.Xml.XmlDocument.dll vs.file.ngenArchitecture=all
-    file source=$(OutputPath)\Vsix\CompilerExtension\System.Xml.XPath.dll vs.file.ngenArchitecture=all
-    file source=$(OutputPath)\Vsix\CompilerExtension\System.Xml.XPath.XDocument.dll vs.file.ngenArchitecture=all
-
-folder InstallDir:\MSBuild\15.0\Bin\amd64
+folder InstallDir:\MSBuild\15.0\Bin\Roslyn
```
Comment (TyOverby) [inline src/Setup/DevDivVsix/CompilersPackage/Microsoft.CodeAnalysis.Compilers.swr]: Does this preserve `ngenArchitecture`?

Comment (OmarTawfik) [inline src/Setup/DevDivVsix/CompilersPackage/Microsoft.CodeAnalysis.Compilers.swr]: @TyOverby  what do you mean? the attribute `vs.file.ngenArchitecture=all` is specified on files in both folders.

**Comment by jaredpar** [comment]
@OmarTawfik the windows_debug_unit32_prtest failures don't appear to be a part of your change.  There was a bad VM that caused 20,000 failures this change hit.  It's been taken out of rotation. 

There does appear to be one real failure though:

> Microsoft.CodeAnalysis.BuildTasks.UnitTests.IntegrationTests.ReportAnalyzerMSBuild

**Comment by jaredpar** [comment]
Hmmm, I think this change exposed a flaw in that tests.  Or possibly the build output refactoring. 

It looks to be failing because it can't find vbc in the same directory.  I'm guessing before your change the fall back logic was **incorrectly** grabbing vbc from the MSBuild directory.  I say incorrectly because the intent of this code was definitely not to test the old compiler but the new one. 

This test itself though seems rather needlessly complex.  Lets sync up in the AM with @agocke and see if we can simplify this test any.

**Comment by OmarTawfik** [comment]
@jaredpar @agocke it is because csc/vbc are not copied to the output directory of MSBuildTaskTests. I'm fixing that now.

**Comment by OmarTawfik** [comment]
@dotnet/roslyn-compiler @TyOverby @agocke  can I get another sign off?

File: src/Compilers/Core/MSBuildTask/InteractiveCompiler.cs [hunk 3]
```diff
@@ -204,16 +204,11 @@ protected override string GenerateCommandLineCommands()
         /// </summary>
         protected override string GenerateFullPathToTool()
         {
-            string pathToTool = ToolLocationHelper.GetPathToBuildToolsFile(ToolName, ToolLocationHelper.CurrentToolsVersion);
+            var pathToTool = Utilities.GenerateFullPathToTool(ToolName);
```
Comment (agocke) [inline src/Compilers/Core/MSBuildTask/InteractiveCompiler.cs]: Maybe create a helper for this method to share between csi and ManagedCompiler?

Comment (OmarTawfik) [inline src/Compilers/Core/MSBuildTask/InteractiveCompiler.cs]: @agocke all code was refactored to `Utilities.GenerateFullPathToTool()`. Having a new base type for both of them just to have one call to`Log.LogErrorWithCodeFromResources()` seems excessive to me. Do you disagree?

**Comment by OmarTawfik** [comment]
Rebasing on dev15-rc2 branch.

**Comment by OmarTawfik** [comment]
Closing as it is being replaced with #15334

**Comment by jaredpar** [comment]
@MattGertz can we get ask mode approval for this PR?  This is the coordinated change between MSBuild and Roslyn to move the Roslyn bits into a sub-directory of MSBuild.  This separation is necessary change due to the removal of MSBuild and Roslyn from the GAC.

**Comment by nguerrera** [comment]
Just making sure the feedback to ensure that satellite assemblies get moved (on the other PR) was retained after this delay. Make sure to move the satellites in VSO when this gets inserted.

**Comment by MattGertz** [comment]
Yes, please (RC.3).

**Comment by OmarTawfik** [comment]
Thank you. Merging now.

File: src/Compilers/Core/MSBuildTask/Utilities.cs [hunk 4]
```diff
@@ -149,5 +151,82 @@ internal static string GetLocation(Assembly assembly)
...
+
+            var buildTask = typeof(Utilities).GetTypeInfo().Assembly;
+            var inGac = (bool?)typeof(Assembly)
+                .GetTypeInfo()
+                .GetDeclaredProperty("GlobalAssemblyCache")
+                ?.GetMethod.Invoke(buildTask, parameters: null);
+
+            if (inGac != true)
+            {
+                var codeBase = (string)typeof(Assembly)
```
Comment (tmat) [inline src/Compilers/Core/MSBuildTask/Utilities.cs]: This is too complicated. You can get the assembly location using ```Assembly.ManifestModule.FullyQualifiedName```

File: src/Compilers/Core/MSBuildTask/Utilities.cs [hunk 5]
```diff
@@ -149,5 +151,82 @@ internal static string GetLocation(Assembly assembly)
...
+                    var uri = new Uri(codeBase);
+
+                    string assemblyPath = null;
+                    if (uri.IsFile)
+                    {
+                        assemblyPath = uri.LocalPath;
+                    }
+                    else
+                    {
+                        var callingAssembly = (Assembly)typeof(Assembly)
```
Comment (tmat) [inline src/Compilers/Core/MSBuildTask/Utilities.cs]: No need to call GetCallingAssembly. Just use ```typeof(Utilities).GetTypeInfo().Assembly```

**Pull request merged**